### ISSUE TITLE
Omitting function body when overriding @:coreApi classes.

### ISSUE THREAD

**Issue by nadako** [HaxeFoundation/haxe#3285]
While working on splitting `haxe.io` stuff, I once again came to think that it would be useful to be able to omit function body for non-extern @:coreApi classes which would mean that function body is copied from the common class.

For example:

Comon `haxe.io.Input`

``` haxe
class Input {
    public function readByte() : Int {
        throw "Not implemented";
    }

    public function readInt8() : Int {
        var n = readByte();
        if( n >= 128 )
            return n - 256;
        return n;
    }
}
```

Target-specific `haxe.io.Input`

``` haxe
@:coreApi class Input {
    public function readByte() : Int {
        return someImplementation();
    }

    public function readInt8() : Int; // use the same algorithm
}
```

**Comment by Simn** [comment]
I have implemented something on this branch: https://github.com/HaxeFoundation/haxe/tree/ast_source

It has two components:

1. `@:astSource`: Adding this to a method causes the compiler to add the method expression as argument. It becomes `@:astSource({ the expression })`.
2. `@:useAstSource dot.path.To.field`: This inserts the `@:astSource` expression of `To.field` in its place. Additionally, if `@:coreApi` is added (`@:useAstSource @:coreApi dot.path.To.field`) then the core API class field is used instead.

With this we can express @nadako's example like so:

```haxe
@:coreApi
class Input {
	public function readByte():Int {
		return 0;
	}

	public function readInt8():Int @:useAstSource @:coreApi haxe.io.Input.readInt8;
}
```

Obviously this could be abbreviated further to make it more elegant.

**Comment by nadako** [comment]
I wonder to what extent this is actually useful, since we still have to manually provide everything that is required for resolving identifiers in copied untyped AST (imports/usings/fields/etc), I'm a bit concerned that this is too error-prone to be a proper language feature.

**Comment by Simn** [comment]
Users certainly have to be aware, but then the same is true for macros that return expressions that are retrieved from external sources.

**Comment by ncannasse** [comment]
I don't like it that much either, that requires to be able to know in advance which AST will have to be reused, which breaks the separation of implementation between the base api and the platform-specific one.

**Comment by Simn** [comment]
You [suggested exactly that](https://github.com/HaxeFoundation/haxe/pull/3327#issuecomment-57934669) and I've implemented it accordingly.

**Comment by ncannasse** [comment]
My suggestion was to have this done automatically, maybe it's the case?

**Comment by Simn** [comment]
Not yet, but to do "something" automatically you have to implement "something" first. Also I'm not sure how to approach this, i.e. how to detect which classes are core classes. This is especially true for your ideas in #4066.

**Comment by ncannasse** [comment]
look at Typeload.load_core_class: we're loading core classes in a separate core context, which have -D coreApi defined

**Comment by Simn** [comment]
Yes I figured that out after mentioning it and added it accordingly. I'm a bit concerned that it's for the most part a pointless waste of memory, but then again it's hardly going to make any difference in the big picture.
